mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/TypeMemberMessager.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.mapper.processor;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+
+/**
+ * Handles warning and error messages for a processed method or field.
+ *
+ * <p>We don't issue those messages on the {@link ExecutableElement} directly, because that can lead
+ * to imprecise compiler messages if the member is inherited from an already compiled parent type.
+ * Consider the following situation:
+ *
+ * <pre>
+ *   interface BaseDao {
+ *     &#64;Select
+ *     void select();
+ *   }
+ *   &#64;Dao
+ *   interface ConcreteDao extends BaseDao {}
+ * </pre>
+ *
+ * If {@code BaseDao} belongs to a JAR dependency, it is already compiled and the warning or error
+ * message can't reference a file or line number, it doesn't even mention {@code ConcreteDao}.
+ *
+ * <p>The goal of this class is to detect those cases, and issue the message on {@code ConcreteDao}
+ * instead.
+ */
+public class TypeMemberMessager {
+
+  private final DecoratedMessager delegate;
+
+  // The element where we'll issue the message
+  private final Element targetElement;
+  // Additional location information that will get prepended to every message
+  private final String locationInfo;
+
+  /**
+   * @param processedType the type that we are currently processing ({@code ConcreteDao} in the
+   *     example above).
+   */
+  public TypeMemberMessager(
+      @NonNull Element member,
+      @NonNull TypeElement processedType,
+      @NonNull ProcessorContext context) {
+
+    this.delegate = context.getMessager();
+
+    // The type that declares the method (BaseDao in the example above). That's the one that might
+    // be already compiled.
+    Element declaringType = member.getEnclosingElement();
+    assert declaringType instanceof TypeElement : "member must be a field or method";
+
+    if (processedType.equals(declaringType) || isSourceFile((TypeElement) declaringType)) {
```

Comment:
Indeed, a weakness of the current design is that it relies on us using `TypeMemberMessager` instead of the default messager when necessary.

Maybe I could instead have all methods in `DecoratedMessager` receive `Element targetElement, TypeElement processedClass`, and do the detection every time.

I'll take a shot at it.

Reply:
Done, this looks much cleaner indeed. LMK what you think.